trappy/ftrace.py

```diff
@@ -97,13 +97,59 @@ def _check_trace_cache(self, params):
             shutil.rmtree(cache_path)
             return False
 
+        self.cache_window_differs = False
+
         # Check if cache can be used with given parameters
-        # Convert to a json string for comparison
+        # Convert to a json string for quick comparison
         if json.dumps(cache_info["params"]) != json.dumps(params):
-            warnstr = "Cached trace parameters differ from those given, invalidating cache."
-            warnings.warn(warnstr)
-            shutil.rmtree(cache_path)
-            return False
+            # Window checks follow these rules:
+            # - Timestamp is not in window if lower than *either* window lower bound
+            # - Timestamp is not in window if greater than *either* window upper bound
+
+            # We'll build the maximum cache window and see if it's included
+            # in the cached window
+            window = params["window"]
+            abs_window = params["abs_window"]
+            max_window = [0, None]
+
+            max_window[0] = max(window[0], abs_window[0] - self.basetime)
+
+            if (window[1] is not None) and (abs_window[1] is not None):
+                max_window[1] = max(window[1], abs_window[1] - self.basetime)
+            elif window[1] is not None:
+                max_window[1] = window[1]
+            elif abs_window[1] is not None:
+                max_window[1] = abs_window[1]
+
+            # Set windows to cache windows for inclusion check
+            self.window = cache_info["params"]["window"]
+            self.abs_window = cache_info["params"]["abs_window"]
+
+            # Add basetime for timestamp check
+            max_window[0] += self.basetime
+            before_window = self.__before_window(max_window[0])
+
+            if max_window[1]:
+                max_window[1] += self.basetime
+                after_window = self.__after_window(max_window[1])
+            else:
+                after_window = self.window[1] or not self.abs_window[1]
+
+            self.max_window = max_window
+
+            # Set windows to parameter windows
+            self.window = window
+            self.abs_window = abs_window
+
+            if before_window or after_window:
+                warnstr = "Cached trace parameters incompatible " + \
+                          "with those given, invalidating cache."
+                warnings.warn(warnstr)
+                shutil.rmtree(cache_path)
+                return False
+
+            self.cache_window_differs = True
```

Comment:
Since we have had correctness issues with this stuff in the past I would not really be comfortable with all this window checking unless there is really good test coverage.. if you have time to add that, then let's go ahead, but otherwise I am more than happy to live with overly-aggressive cache invalidation in return for a little bit more confidence about correctness. After all I don't think the window param even gets used very often..

Reply:
I don't use the window parameters that much myself either - I manually trim the dataframes if I have to. Thing is, if someone does use them, this makes quite a big difference in terms of parsing time.